diff --git a/src/physfs_platforms.h b/src/physfs_platforms.h
--- a/src/physfs_platforms.h
+++ b/src/physfs_platforms.h
@@ -86,6 +86,10 @@
 #elif defined(unix) || defined(__unix__)
 #  define PHYSFS_PLATFORM_UNIX 1
 #  define PHYSFS_PLATFORM_POSIX 1
+#elif defined(PSP)
+#  define PHYSFS_NO_CDROM_SUPPORT 1
+#  define PHYSFS_PLATFORM_UNIX 1
+#  define PHYSFS_PLATFORM_POSIX 1
 #elif defined(__wii__) || defined(__gamecube__)
 #  define PHYSFS_PLATFORM_OGC 1
 #  define PHYSFS_NO_CDROM_SUPPORT 1 // TODO
