diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -145,7 +145,7 @@ if(lto_supported)
   if(PROJECT_IS_TOP_LEVEL AND BUILD_FUZZER)
     set_target_properties(unarr PROPERTIES INTERPROCEDURAL_OPTIMIZATION FALSE)
   else()
-    set_target_properties(unarr PROPERTIES INTERPROCEDURAL_OPTIMIZATION TRUE)
+    set_target_properties(unarr PROPERTIES INTERPROCEDURAL_OPTIMIZATION FALSE)
   endif()
 else()
   message(WARNING "IPO is not supported: ${lto_diagnostic}")
diff --git a/common/unarr-imp.h b/common/unarr-imp.h
--- a/common/unarr-imp.h
+++ b/common/unarr-imp.h
@@ -4,6 +4,7 @@
 #ifndef common_unarr_imp_h
 #define common_unarr_imp_h
 
+#include <sys/_stdint.h>
 #include "unarr.h"
 #include "allocator.h"
 
