diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -146,2 +146,7 @@
-    list(APPEND THIRD_PARTY_LIBS Threads::Threads)
-    list(APPEND PC_LIBS_PRIV -pthread)
+    if (PSP)
+      list(APPEND THIRD_PARTY_LIBS pthread)
+      list(APPEND PC_LIBS_PRIV -lpthread)
+    else()
+      list(APPEND THIRD_PARTY_LIBS Threads::Threads)
+      list(APPEND PC_LIBS_PRIV -pthread)
+    endif ()
@@ -746,1 +751,1 @@
-if (UNIX OR MINGW OR VITA)
+if (UNIX OR MINGW OR VITA OR PSP)
