diff --git a/library/net_sockets.c b/library/net_sockets.c
--- a/library/net_sockets.c
+++ b/library/net_sockets.c
@@ -9,6 +9,6 @@
 
 #if defined(MBEDTLS_NET_C)
 
-#if !defined(MBEDTLS_PLATFORM_IS_UNIXLIKE) && !defined(_WIN32)
+#if !defined(MBEDTLS_PLATFORM_IS_UNIXLIKE) && !defined(_WIN32) && !defined(PSP)
 #error "This module only works on Unix and Windows, see MBEDTLS_NET_C in mbedtls_config.h"
 #endif
