diff --git a/sharedbook.c b/sharedbook.c
--- a/sharedbook.c
+++ b/sharedbook.c
@@ -230,8 +230,8 @@ ogg_int32_t *_book_unquantize(const static_codebook *b,int n,int *sparsemap,
           int lastpoint=0;
           int indexdiv=1;
           for(k=0;k<b->dim;k++){
             int index= (j/indexdiv)%quantvals;
-            int point=0;
-            int val=VFLOAT_MULTI(delta,delpoint,
+            ogg_int32_t point=0;
+            ogg_int32_t val=VFLOAT_MULTI(delta,delpoint,
                                  abs(b->quantlist[index]),&point);
 
             val=VFLOAT_ADD(mindel,minpoint,val,point,&point);
@@ -264,8 +264,8 @@ ogg_int32_t *_book_unquantize(const static_codebook *b,int n,int *sparsemap,
           int lastpoint=0;
 
           for(k=0;k<b->dim;k++){
-            int point=0;
-            int val=VFLOAT_MULTI(delta,delpoint,
+            ogg_int32_t point=0;
+            ogg_int32_t val=VFLOAT_MULTI(delta,delpoint,
                                  abs(b->quantlist[j*b->dim+k]),&point);
 
             val=VFLOAT_ADD(mindel,minpoint,val,point,&point);
