diff --git a/makefile b/makefile
--- a/makefile
+++ b/makefile
@@ -5,13 +5,13 @@
 # 2024.08.19: -march=native isn't recognized on some platforms such as RISCV64.
 # Thus we removed it. Clang ARM users can add -march=armv8-a+crypto to enable
 # ARM NEON crypto.
-CXX=c++
-CXXFLAGS=-O2 -std=c++11 -Wno-logical-op-parentheses -Wno-switch -Wno-dangling-else
-LIBFLAGS=-fPIC
-DEFINES=-D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DRAR_SMP
-STRIP=strip
-AR=ar
-LDFLAGS=-pthread
+CXX=psp-g++
+CXXFLAGS=-O2 -G0 -std=gnu++11 -fno-rtti -Wall -Wno-unused-variable -Wno-unused-but-set-variable
+LIBFLAGS=
+DEFINES=-D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DLITTLE_ENDIAN
+STRIP=psp-strip
+AR=psp-ar
+LDFLAGS=
 DESTDIR=/usr
 
 ##########################
@@ -62,7 +62,6 @@
 lib:	CXXFLAGS+=$(LIBFLAGS)
 lib:	$(OBJECTS) $(LIB_OBJ)
 	@rm -f libunrar.*
-	$(LINK) -shared -o libunrar.so $(LDFLAGS) $(OBJECTS) $(LIB_OBJ)
 	$(AR) rcs libunrar.a $(OBJECTS) $(LIB_OBJ)
 
 install-unrar:
diff --git a/os.hpp b/os.hpp
--- a/os.hpp
+++ b/os.hpp
@@ -128,7 +128,7 @@
 #ifdef _APPLE
   #include <sys/sysctl.h>
 #endif
-#ifndef SFX_MODULE
+#if !defined(SFX_MODULE) && !defined(__PSP__)
     #include <sys/statvfs.h>
 #endif
 #include <pwd.h>
@@ -223,7 +223,9 @@
 #endif
 
 // Unlike Apple x64, utimensat shall be available in all Apple M1 systems.
-#if _POSIX_C_SOURCE >= 200809L || defined(__APPLE__) && defined(__arm64__)
+#if defined(__PSP__)
+  extern "C" int utime(const char *path, const struct utimbuf *times);
+#elif _POSIX_C_SOURCE >= 200809L || defined(__APPLE__) && defined(__arm64__)
   #define UNIX_TIME_NS // Nanosecond time precision in Unix.
 #endif
 
diff --git a/filefn.cpp b/filefn.cpp
--- a/filefn.cpp
+++ b/filefn.cpp
@@ -158,6 +158,10 @@
     return INT32TO64(uiUserFree.u.HighPart,uiUserFree.u.LowPart);
   return 0;
 #elif defined(_UNIX)
+#ifdef __PSP__
+  (void)Name;
+  return 0;
+#else
   std::wstring Root;
   GetPathWithSep(Name,Root);
   std::string RootA;
@@ -168,6 +172,7 @@
   int64 FreeSize=sfs.f_bsize;
   FreeSize=FreeSize*sfs.f_bavail;
   return FreeSize;
+#endif
 #else
   return 0;
 #endif
