diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -139,6 +139,8 @@
   if (CMAKE_USE_PTHREADS_INIT)
     add_compile_definitions(HAVE_PTHREAD)
-    list(APPEND THIRD_PARTY_LIBS Threads::Threads)
+    if (NOT PSP)
+      list(APPEND THIRD_PARTY_LIBS Threads::Threads)
+    endif ()
     list(APPEND PC_LIBS_PRIV -pthread)
   endif ()
 endif ()
@@ -746,7 +748,7 @@
-if (UNIX OR MINGW OR VITA)
+if (UNIX OR MINGW OR VITA OR PSP)
   # Make symbols link locally
   include (CheckCXXCompilerFlag)
   CHECK_CXX_COMPILER_FLAG(-Bsymbolic-functions CXX_SUPPORTS_FLAG_BSYMB_FUNCS)
