diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -139,8 +139,13 @@
   find_package(Threads)
   if (CMAKE_USE_PTHREADS_INIT)
     add_compile_definitions(HAVE_PTHREAD)
-    list(APPEND THIRD_PARTY_LIBS Threads::Threads)
-    list(APPEND PC_LIBS_PRIV -pthread)
+    if (PSP)
+      list(APPEND THIRD_PARTY_LIBS pthread)
+      list(APPEND PC_LIBS_PRIV -lpthread)
+    else()
+      list(APPEND THIRD_PARTY_LIBS Threads::Threads)
+      list(APPEND PC_LIBS_PRIV -pthread)
+    endif ()
   endif ()
 endif ()
@@ -746,4 +751,4 @@
-if (UNIX OR MINGW OR VITA)
+if (UNIX OR MINGW OR VITA OR PSP)
   # Make symbols link locally
   include (CheckCXXCompilerFlag)
   CHECK_CXX_COMPILER_FLAG(-Bsymbolic-functions CXX_SUPPORTS_FLAG_BSYMB_FUNCS)
